### PR TITLE
app-crypt/tpm2-pkcs11: New package

### PR DESCRIPTION
**Pull request by salahcoronya**
PKCS #11 is a Public-Key Cryptography Standard that defines a standard method
to access cryptographic services from tokens/ devices such as hardware security
modules (HSM), smart cards, etc. This package intends to use a TPM2 device as
the cryptographic token.

Package-Manager: Portage-2.3.89, Repoman-2.3.20
Signed-off-by: Salah Coronya <salah.coronya@gmail.com>

### CODE REVIEW

File: app-crypt/tpm2-pkcs11/tpm2-pkcs11-1.2.0.ebuild [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+LICENSE="BSD-2"
+SLOT="0"
+KEYWORDS="~amd64"
+
+# Test requires the IBM TPM simulator that's not in portage
+RESTRICT="test"
+
+RDEPEND="app-crypt/p11-kit
+	>=app-crypt/tpm2-abrmd-2.1.0
```
Comment (juippis) [inline app-crypt/tpm2-pkcs11/tpm2-pkcs11-1.2.0.ebuild:24]: Always satisfied in ::gentoo tree, version restriction can be lifted off.

File: app-crypt/tpm2-pkcs11/tpm2-pkcs11-1.2.0.ebuild [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+	>=app-crypt/tpm2-tss-2.3.0
+	dev-db/sqlite
+	>=dev-libs/openssl-1.0.2
```
Comment (juippis) [inline app-crypt/tpm2-pkcs11/tpm2-pkcs11-1.2.0.ebuild:27]: Version restriction not needed here either.

File: app-crypt/tpm2-pkcs11/tpm2-pkcs11-1.2.0.ebuild [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+SLOT="0"
+KEYWORDS="~amd64"
+
+# Test requires the IBM TPM simulator that's not in portage
+RESTRICT="test"
+
+RDEPEND="app-crypt/p11-kit
+	>=app-crypt/tpm2-abrmd-2.1.0
+	>=app-crypt/tpm2-tss-2.3.0
+	dev-db/sqlite
```
Comment (juippis) [inline app-crypt/tpm2-pkcs11/tpm2-pkcs11-1.2.0.ebuild:26]: Add a slot operator here.

File: app-crypt/tpm2-pkcs11/tpm2-pkcs11-1.2.0.ebuild [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+	>=dev-libs/openssl-1.0.2
+	dev-python/pyyaml[${PYTHON_USEDEP}]
+	dev-python/cryptography[${PYTHON_USEDEP}]
+	dev-python/pyasn1[${PYTHON_USEDEP}]
+	dev-python/pyasn1-modules[${PYTHON_USEDEP}]"
+
+# Build-time dependencies that need to be binary compatible with the system
+# being built (CHOST). These include libraries that we link against.
+# The below is valid if the same run-time depends are required to compile.
```
Comment (juippis) [inline app-crypt/tpm2-pkcs11/tpm2-pkcs11-1.2.0.ebuild:35]: Can be removed.

File: app-crypt/tpm2-pkcs11/tpm2-pkcs11-1.2.0.ebuild [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+DEPEND="${RDEPEND}"
+
+# Build-time dependencies that are executed during the emerge process, and
+# only need to be present in the native build system (CBUILD). Example:
```
Comment (juippis) [inline app-crypt/tpm2-pkcs11/tpm2-pkcs11-1.2.0.ebuild:39]: Can be removed.

File: app-crypt/tpm2-pkcs11/tpm2-pkcs11-1.2.0.ebuild [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+BDEPEND=">=sys-devel/autoconf-archive-2018.03.13
```
Comment (juippis) [inline app-crypt/tpm2-pkcs11/tpm2-pkcs11-1.2.0.ebuild:40]: Version restriction can be removed.

File: app-crypt/tpm2-pkcs11/tpm2-pkcs11-1.2.0.ebuild [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+	virtual/pkgconfig"
+
+REQUIRED_USE="( ${PYTHON_REQUIRED_USE} )"
+
+src_configure()
+{
+	econf
+}
```
Comment (juippis) [inline app-crypt/tpm2-pkcs11/tpm2-pkcs11-1.2.0.ebuild:48]: Default phase, can be removed.

File: app-crypt/tpm2-pkcs11/tpm2-pkcs11-1.2.0.ebuild [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+
+src_compile()
+{
```
Comment (juippis) [inline app-crypt/tpm2-pkcs11/tpm2-pkcs11-1.2.0.ebuild:51]: Single-line this, `src_compile() {`

File: app-crypt/tpm2-pkcs11/tpm2-pkcs11-1.2.0.ebuild [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+	emake
+	cd "${S}/tools" && BUILD_DIR="${S}/tools" python_foreach_impl distutils-r1_python_compile
```
Comment (juippis) [inline app-crypt/tpm2-pkcs11/tpm2-pkcs11-1.2.0.ebuild:53]: `cd` command need to be appended with `|| die`, so you'll have to split this.

File: app-crypt/tpm2-pkcs11/tpm2-pkcs11-1.2.0.ebuild [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+}
+
+src_install()
+{
```
Comment (juippis) [inline app-crypt/tpm2-pkcs11/tpm2-pkcs11-1.2.0.ebuild:57]: `src_install() {`

File: app-crypt/tpm2-pkcs11/tpm2-pkcs11-1.2.0.ebuild [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+	emake DESTDIR="${D}" install
+	einstalldocs
```
Comment (juippis) [inline app-crypt/tpm2-pkcs11/tpm2-pkcs11-1.2.0.ebuild:59]: `default`

File: app-crypt/tpm2-pkcs11/tpm2-pkcs11-1.2.0.ebuild [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+	cd "${S}/tools" && BUILD_DIR="${S}/tools" python_foreach_impl distutils-r1_python_install
```
Comment (juippis) [inline app-crypt/tpm2-pkcs11/tpm2-pkcs11-1.2.0.ebuild:60]: Add `|| die` and multi-line this properly.

File: app-crypt/tpm2-pkcs11/tpm2-pkcs11-1.2.0.ebuild (hunk 1 above)
Comment (salahcoronya) [reply app-crypt/tpm2-pkcs11/tpm2-pkcs11-1.2.0.ebuild]: Done

File: app-crypt/tpm2-pkcs11/tpm2-pkcs11-1.2.0.ebuild (hunk 4 above)
Comment (salahcoronya) [reply app-crypt/tpm2-pkcs11/tpm2-pkcs11-1.2.0.ebuild]: Done

File: app-crypt/tpm2-pkcs11/tpm2-pkcs11-1.2.0.ebuild (hunk 8 above)
Comment (salahcoronya) [reply app-crypt/tpm2-pkcs11/tpm2-pkcs11-1.2.0.ebuild]: Done

File: app-crypt/tpm2-pkcs11/tpm2-pkcs11-1.2.0.ebuild (hunk 6 above)
Comment (salahcoronya) [reply app-crypt/tpm2-pkcs11/tpm2-pkcs11-1.2.0.ebuild]: Done

File: app-crypt/tpm2-pkcs11/tpm2-pkcs11-1.2.0.ebuild (hunk 7 above)
Comment (salahcoronya) [reply app-crypt/tpm2-pkcs11/tpm2-pkcs11-1.2.0.ebuild]: Doesn't see to work for me. the distutils-r1 eclass doesn't run ./configure with explicilty calling the default version.

File: app-crypt/tpm2-pkcs11/tpm2-pkcs11-1.2.0.ebuild (hunk 9 above)
Comment (salahcoronya) [reply app-crypt/tpm2-pkcs11/tpm2-pkcs11-1.2.0.ebuild]: Split

File: app-crypt/tpm2-pkcs11/tpm2-pkcs11-1.2.0.ebuild (hunk 10 above)
Comment (salahcoronya) [reply app-crypt/tpm2-pkcs11/tpm2-pkcs11-1.2.0.ebuild]: Fixed

File: app-crypt/tpm2-pkcs11/tpm2-pkcs11-1.2.0.ebuild (hunk 11 above)
Comment (salahcoronya) [reply app-crypt/tpm2-pkcs11/tpm2-pkcs11-1.2.0.ebuild]: Done

File: app-crypt/tpm2-pkcs11/tpm2-pkcs11-1.2.0.ebuild (hunk 12 above)
Comment (salahcoronya) [reply app-crypt/tpm2-pkcs11/tpm2-pkcs11-1.2.0.ebuild]: Added and split

File: app-crypt/tpm2-pkcs11/tpm2-pkcs11-1.2.0.ebuild (hunk 2 above)
Comment (salahcoronya) [reply app-crypt/tpm2-pkcs11/tpm2-pkcs11-1.2.0.ebuild]: Removed

File: app-crypt/tpm2-pkcs11/tpm2-pkcs11-1.2.0.ebuild (hunk 3 above)
Comment (salahcoronya) [reply app-crypt/tpm2-pkcs11/tpm2-pkcs11-1.2.0.ebuild]: Added

File: app-crypt/tpm2-pkcs11/tpm2-pkcs11-1.2.0.ebuild [hunk 13]
```diff
@@ -0,0 +1,54 @@
...
+SLOT="0"
+KEYWORDS="~amd64"
+
+# Test requires the IBM TPM simulator that's not in portage
+RESTRICT="test"
+
+RDEPEND="app-crypt/p11-kit
+	app-crypt/tpm2-abrmd
+	app-crypt/tpm2-tss
+	dev-db/sqlite:=
```
Comment (juippis) [inline app-crypt/tpm2-pkcs11/tpm2-pkcs11-1.2.0.ebuild:26]: sqlite doesn't have a subslot, so binder is useless here. You want to depend on a slot, so either `:3` or `:*`. Drop the `=`

File: app-crypt/tpm2-pkcs11/tpm2-pkcs11-1.2.0.ebuild [hunk 14]
```diff
@@ -0,0 +1,54 @@
...
+	dev-libs/openssl
+	dev-python/pyyaml[${PYTHON_USEDEP}]
+	dev-python/cryptography[${PYTHON_USEDEP}]
+	dev-python/pyasn1[${PYTHON_USEDEP}]
+	dev-python/pyasn1-modules[${PYTHON_USEDEP}]"
```
Comment (juippis) [inline app-crypt/tpm2-pkcs11/tpm2-pkcs11-1.2.0.ebuild:31]: Please sort in alphabetical order.

File: app-crypt/tpm2-pkcs11/tpm2-pkcs11-1.2.0.ebuild [hunk 15]
```diff
@@ -0,0 +1,54 @@
...
+
+DEPEND="${RDEPEND}"
+BDEPEND="sys-devel/autoconf-archive
+	virtual/pkgconfig"
+
+REQUIRED_USE="( ${PYTHON_REQUIRED_USE} )"
```
Comment (juippis) [inline app-crypt/tpm2-pkcs11/tpm2-pkcs11-1.2.0.ebuild:37]: This is not needed with *distutils-r1.eclass*.

File: app-crypt/tpm2-pkcs11/tpm2-pkcs11-1.2.0.ebuild [hunk 16]
```diff
@@ -0,0 +1,54 @@
...
+	default
+	cd "${S}/tools" || die
+	BUILD_DIR="${S}/tools" python_foreach_impl distutils-r1_python_compile
+}
+
+src_install() {
+	default
+	cd "${S}/tools" || die
+	BUILD_DIR="${S}/tools" python_foreach_impl distutils-r1_python_install
+	dobin "${S}/tools/tpm2_ptool" || die
```
Comment (juippis) [inline app-crypt/tpm2-pkcs11/tpm2-pkcs11-1.2.0.ebuild:53]: `dobin` is an ebuild helper, so no `die` needed here. I know, takes some time to learn :)

File: app-crypt/tpm2-pkcs11/tpm2-pkcs11-1.2.0.ebuild (hunk 5 above)
Comment (salahcoronya) [reply app-crypt/tpm2-pkcs11/tpm2-pkcs11-1.2.0.ebuild]: Done

File: app-crypt/tpm2-pkcs11/tpm2-pkcs11-1.2.0.ebuild (hunk 13 above)
Comment (salahcoronya) [reply app-crypt/tpm2-pkcs11/tpm2-pkcs11-1.2.0.ebuild]: ok I fixed it properly now

File: app-crypt/tpm2-pkcs11/tpm2-pkcs11-1.2.0.ebuild (hunk 14 above)
Comment (salahcoronya) [reply app-crypt/tpm2-pkcs11/tpm2-pkcs11-1.2.0.ebuild]: Sorted

File: app-crypt/tpm2-pkcs11/tpm2-pkcs11-1.2.0.ebuild (hunk 15 above)
Comment (salahcoronya) [reply app-crypt/tpm2-pkcs11/tpm2-pkcs11-1.2.0.ebuild]: ok, I dropped the REQUIRED_USE line .

File: app-crypt/tpm2-pkcs11/tpm2-pkcs11-1.2.0.ebuild (hunk 16 above)
Comment (salahcoronya) [reply app-crypt/tpm2-pkcs11/tpm2-pkcs11-1.2.0.ebuild]: removed

**Pull request closed without merging**